editor/src/core/shared/element-path-tree.ts

```diff
@@ -1,140 +1,146 @@
-  trees[rootPathString] = elementPathTree(rootPath, rootPathString, [])
-
-  const childrenPaths = getChildrenPaths(metadata, rootPath, originalPaths)
+): void {
+  addRootToTree(rootPath, trees)
 
-  let children: ElementPathTree[] = []
-  for (const path of childrenPaths) {
+  function addPathToTree(path: ElementPath): void {
     const pathString = EP.toString(path)
-    const subTree = elementPathTree(
-      path,
-      pathString,
-      buildTreeRecursive_MUTATE(path, trees, originalPaths, metadata),
-    )
-    trees[rootPathString].children.push(subTree)
-    children.push(subTree)
-  }
+    const alreadyAdded = pathString in trees
+    if (!alreadyAdded) {
+      const parentPath = EP.parentPath(path)
+      const subTree = elementPathTree(path, pathString, [], [])
+      trees[pathString] = subTree
+      const pathLastPart = EP.lastElementPathForPath(path)
 
-  return children
-}
+      if (pathLastPart?.length === 1) {
+        // If the last part of the path is a single element, it should be added to the inner children.
+        // This would be a path like `a/b/c:d`.
+        trees[EP.toString(parentPath)].innerChildren.push(subTree)
+      } else {
+        // Otherwise this should be added to the props children.
+        trees[EP.toString(parentPath)].propsChildren.push(subTree)
+      }
+    }
+  }
 
-function getChildrenPaths(
-  metadata: ElementInstanceMetadataMap,
-  rootPath: ElementPath,
-  paths: ElementPath[],
-): ElementPath[] {
-  const element = metadata[EP.toString(rootPath)]
+  let workingPaths = [...originalPaths]
+  workingPaths.sort((a, b) => {
+    return EP.fullDepth(a) - EP.fullDepth(b)
+  })
+  for (const workingPath of workingPaths) {
+    addPathToTree(workingPath)
+  }
 
-  // Grab the child elements from the element metadata, if available.
-  let childrenFromElement: ElementPath[] = []
-  if (
-    element != null &&
-    isRight(element.element) &&
-    (isJSXElement(element.element.value) || isJSXFragment(element.element.value)) &&
-    element.element.value.children.length > 0
-  ) {
-    childrenFromElement = element.element.value.children
-      .filter((child) => {
+  for (const elementMetadata of Object.values(metadata)) {
+    forEachRight(elementMetadata.element, (element) => {
+      if (isJSXElementLike(element)) {
+        let elementChildren: Array<JSXElementChild>
         if (isFeatureEnabled('Condensed Navigator Entries')) {
-          // if Data Entries are enabled, we should always show them in the Navigator
-          return true
+          elementChildren = element.children
         } else {
-          return (
-            !isJSXTextBlock(child) &&
-            !isJSExpressionMapOrOtherJavaScript(child) &&
-            !isJSIdentifier(child) &&
-            !isJSPropertyAccess(child) &&
-            !isJSElementAccess(child)
-          )
+          elementChildren = element.children.filter((child) => {
+            switch (child.type) {
```

Comment:
Given the number of cases here, is it worth making this exhaustive?

Reply:
Sure, it only ended up like this as it was translated from the original condition.

Reply:
Fixed in e0d58e345.